<!DOCTYPE html>
<html lang="ja-JP">
<head>
<!-- Global Site Tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-620926-3"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments)};
  gtag('js', new Date());

  gtag('config', 'UA-620926-3');
</script>

<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" href="../style.css">
<link rel="stylesheet" href="../syntax-highlight.css">
<link rel="icon" type="image/png" href="../rurema.png">

<link rel="canonical" href="https://docs.ruby-lang.org/ja/latest/class/Exception.html">

<title>class Exception (Ruby 2.4.0 リファレンスマニュアル)</title>
<meta name="description" content="全ての例外の祖先のクラスです。">
<script src="../script.js"></script>
</head>
<body>


<script type="application/ld+json">
{"@context":"http://schema.org","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","item":"https://docs.ruby-lang.org/ja/latest/doc/index.html","name":"Ruby 2.4.0 リファレンスマニュアル","position":1},{"@type":"ListItem","item":"https://docs.ruby-lang.org/ja/latest/library/index.html","name":"ライブラリ一覧","position":2},{"@type":"ListItem","item":"https://docs.ruby-lang.org/ja/latest/library/_builtin.html","name":"組み込みライブラリ","position":3},{"@type":"ListItem","item":"https://docs.ruby-lang.org/ja/latest/class/Exception.html","name":"Exceptionクラス","position":4}]}
</script>

<header>
  <nav>
    <ol class="inline-breadcrumb-list">
      <li>
        <a href="../doc/index.html">Ruby 2.4.0 リファレンスマニュアル</a>
      </li>
      <li>
        <a href="../library/index.html">ライブラリ一覧</a>
      </li>
      <li>
        <a href="../library/_builtin.html">組み込みライブラリ</a>
      </li>
      <li>
        Exceptionクラス
      </li>
    </ol>
  </nav>
</header>


<h1>class Exception</h1>



  <nav>
    クラスの継承リスト:
    <ol class="inline-ancestors-list">
      
        <li>
          <a href="../class/BasicObject.html">BasicObject</a>
        </li>
        
      
        <li>
          <a href="../class/Kernel.html">Kernel</a>
        </li>
        
      
        <li>
          <a href="../class/Object.html">Object</a>
        </li>
        
      
      <li>
        Exception
      </li>
    </ol>
  </nav>











<h2>要約</h2>
<p>
全ての例外の祖先のクラスです。
</p>



<h2>目次</h2>
<dl>

<dt>特異メソッド</dt>
<dd>
<ul class="class-toc">

<li><a href="#S_EXCEPTION">exception</a></li>

<li><a href="#S_EXCEPTION">new</a></li>

</ul>
</dd>

<dt>インスタンスメソッド</dt>
<dd>
<ul class="class-toc">

<li><a href="#I_--3D--3D">==</a></li>

<li><a href="#I_BACKTRACE">backtrace</a></li>

<li><a href="#I_BACKTRACE_LOCATIONS">backtrace_locations</a></li>

<li><a href="#I_CAUSE">cause</a></li>

<li><a href="#I_EXCEPTION">exception</a></li>

<li><a href="#I_INSPECT">inspect</a></li>

<li><a href="#I_MESSAGE">message</a></li>

<li><a href="#I_SET_BACKTRACE">set_backtrace</a></li>

<li><a href="#I_MESSAGE">to_s</a></li>

</ul>
</dd>

</dl>




</dl>


<h2>特異メソッド</h2>
<dl>

<dt class="method-heading" id="S_EXCEPTION"><code>new(error_message = nil) -&gt; Exception</code><span class="permalink">[<a href="../method/Exception/s/exception.html">permalink</a>][<a href="https://docs.ruby-lang.org/en/2.4.0/Exception.html#method-c-exception">rdoc</a>]</span></dt>
<dt class="method-heading"><code>exception(error_message = nil) -&gt; Exception</code></dt>
<dd class="method-description">
<p>
例外オブジェクトを生成して返します。
</p>
<dl>
<dt class='method-param'>[PARAM] error_message:</dt>
<dd>
エラーメッセージを表す文字列を指定します。このメッセージは属性 <a href="../method/Exception/i/message.html">Exception#message</a> の値になり、デフォルトの例外ハンドラで表示されます。
</dd>
</dl>
<p>
例:
</p>
<pre>
e = Exception.new(&quot;some message&quot;)
p e         # =&gt; #&lt;Exception: some message&gt;
p e.message # =&gt; &quot;some message&quot;
</pre>
<p>
例:
</p>
<pre>
e = Exception.exception(&quot;some message&quot;)
p e         # =&gt; #&lt;Exception: some message&gt;
p e.message # =&gt; &quot;some message&quot;
</pre>
</dd>


</dl>

<h2>インスタンスメソッド</h2>
<dl>

<dt class="method-heading" id="I_--3D--3D"><code>self == other -&gt; bool</code><span class="permalink">[<a href="../method/Exception/i/=3d=3d.html">permalink</a>][<a href="https://docs.ruby-lang.org/en/2.4.0/Exception.html#method-i-3D-3D">rdoc</a>]</span></dt>
<dd class="method-description">
<p>
自身と指定された other のクラスが同じであり、
message と backtrace が == メソッドで比較して等しい場合に true を返します。そうでない場合に false を返します。
</p>
<dl>
<dt class='method-param'>[PARAM] other:</dt>
<dd>
自身と比較したいオブジェクトを指定します。自身と異なるクラスのオブジェクトを指定した場合は
<a href="../method/Exception/i/exception.html">Exception#exception</a> を実行して変換を試みます。
</dd>
</dl>
<p>
例:
</p>
<pre>
require &quot;date&quot;
def check_long_month(month)
  return if Date.new(2000, month, -1).day == 31
  raise &quot;#{month} is not long month&quot;
end

def get_exception
  return begin
    yield
  rescue =&gt; e
    e
  end
end

results = [2, 2, 4].map { |e | get_exception { check_long_month(e) } }
p results.map { |e| e.class }
# =&gt; [RuntimeError, RuntimeError, RuntimeError]
p results.map { |e| e.message }
# =&gt; [&quot;2 is not long month&quot;, &quot;2 is not long month&quot;, &quot;4 is not long month&quot;]

# class, message, backtrace が同一のため true になる
p results[0] == results[1]    # =&gt; true

# class, backtrace が同一だが、message がことなるため false になる
p results[0] == results[2]    # =&gt; false
</pre>
</dd>


<dt class="method-heading" id="I_BACKTRACE"><code>backtrace -&gt; [String]</code><span class="permalink">[<a href="../method/Exception/i/backtrace.html">permalink</a>][<a href="https://docs.ruby-lang.org/en/2.4.0/Exception.html#method-i-backtrace">rdoc</a>]</span></dt>
<dd class="method-description">
<p>
バックトレース情報を返します。
</p>
<p>
デフォルトでは
</p>
<ul>
<li>&quot;#{sourcefile}:#{sourceline}:in `#{method}'&quot;
(メソッド内の場合)</li>
<li>&quot;#{sourcefile}:#{sourceline}&quot;
(トップレベルの場合)</li>
</ul>
<p>
という形式の <a href="../class/String.html">String</a> の配列です。
</p>
<pre>
def methd
  raise
end

begin
  methd
rescue =&gt; e
  p e.backtrace
end

#=&gt; [&quot;filename.rb:2:in `methd'&quot;, &quot;filename.rb:6&quot;]
</pre>
<p>
[SEE_ALSO] <a href="../method/Exception/i/backtrace_locations.html">Exception#backtrace_locations</a>
</p>
</dd>


<dt class="method-heading" id="I_BACKTRACE_LOCATIONS"><code>backtrace_locations -&gt; [Thread::Backtrace::Location]</code><span class="permalink">[<a href="../method/Exception/i/backtrace_locations.html">permalink</a>][<a href="https://docs.ruby-lang.org/en/2.4.0/Exception.html#method-i-backtrace_locations">rdoc</a>]</span></dt>
<dd class="method-description">
<p>
バックトレース情報を返します。<a href="../method/Exception/i/backtrace.html">Exception#backtrace</a>に似ていますが、
<a href="../class/Thread=3a=3aBacktrace=3a=3aLocation.html">Thread::Backtrace::Location</a> の配列を返す点が異なります。
</p>
<p>
現状では <a href="../method/Exception/i/set_backtrace.html">Exception#set_backtrace</a> によって戻り値が変化する事はありません。
</p>
<p>
例: test.rb
</p>
<pre>
require &quot;date&quot;
def check_long_month(month)
  return if Date.new(2000, month, -1).day == 31
  raise &quot;#{month} is not long month&quot;
end

def get_exception
  return begin
    yield
  rescue =&gt; e
    e
  end
end

e = get_exception { check_long_month(2) }
p e.backtrace_locations
# =&gt; [&quot;test.rb:4:in `check_long_month'&quot;, &quot;test.rb:15:in `block in &lt;main&gt;'&quot;, &quot;test.rb:9:in `get_exception'&quot;, &quot;test.rb:15:in `&lt;main&gt;'&quot;]
</pre>
<p>
[SEE_ALSO] <a href="../method/Exception/i/backtrace.html">Exception#backtrace</a>
</p>
</dd>


<dt class="method-heading" id="I_CAUSE"><code>cause -&gt; Exception | nil</code><span class="permalink">[<a href="../method/Exception/i/cause.html">permalink</a>][<a href="https://docs.ruby-lang.org/en/2.4.0/Exception.html#method-i-cause">rdoc</a>]</span></dt>
<dd class="method-description">
<p>
self の前の例外(self が rescue 節や ensure 節の中で発生した例外の場合、その前に発生していた元々の例外)を返します。存在しない場合は nil を返します。
</p>
<p>
例:
</p>
<pre>
begin
  begin
    raise &quot;inner&quot;
  rescue
    raise &quot;outer&quot;
  end
rescue
  p $!       # =&gt; #&lt;RuntimeError: outer&gt;
  p $!.cause # =&gt; #&lt;RuntimeError: inner&gt;
end
</pre>
</dd>


<dt class="method-heading" id="I_EXCEPTION"><code>exception -&gt; self</code><span class="permalink">[<a href="../method/Exception/i/exception.html">permalink</a>][<a href="https://docs.ruby-lang.org/en/2.4.0/Exception.html#method-i-exception">rdoc</a>]</span></dt>
<dt class="method-heading"><code>exception(error_message) -&gt; Exception</code></dt>
<dd class="method-description">
<p>
引数を指定しない場合は self を返します。引数を指定した場合 自身のコピーを生成し <a href="../method/Exception/i/message.html">Exception#message</a> 属性を error_message にして返します。
</p>
<p>
<a href="../method/Kernel/m/raise.html">Kernel.#raise</a> は、実質的に、例外オブジェクトの exception
メソッドの呼び出しです。
</p>
<dl>
<dt class='method-param'>[PARAM] error_message:</dt>
<dd>
エラーメッセージを表す文字列を指定します。
</dd>
</dl>
<pre>
begin
 ...        # 何か処理
rescue =&gt; e
 raise e.exception(&quot;an error occurs during hogehoge process&quot;)  # 詳しいエラーメッセージ
end
</pre>
</dd>


<dt class="method-heading" id="I_INSPECT"><code>inspect -&gt; String</code><span class="permalink">[<a href="../method/Exception/i/inspect.html">permalink</a>][<a href="https://docs.ruby-lang.org/en/2.4.0/Exception.html#method-i-inspect">rdoc</a>]</span></dt>
<dd class="method-description">
<p>
self のクラス名と message を文字列にして返します。
</p>
<p>
例:
</p>
<pre>
begin
  raise &quot;exception&quot;
rescue
  p $!.inspect # =&gt; &quot;#&lt;RuntimeError: exception&gt;&quot;
end
</pre>
</dd>


<dt class="method-heading" id="I_MESSAGE"><code>message -&gt; String</code><span class="permalink">[<a href="../method/Exception/i/message.html">permalink</a>][<a href="https://docs.ruby-lang.org/en/2.4.0/Exception.html#method-i-message">rdoc</a>]</span></dt>
<dt class="method-heading"><code>to_s -&gt; String</code></dt>
<dd class="method-description">
<p>
エラーメッセージをあらわす文字列を返します。
</p>
<pre>
begin
  1 + nil
rescue =&gt; e
  p e.message   #=&gt;  &quot;nil can't be coerced into Fixnum&quot;
end
</pre>
</dd>


<dt class="method-heading" id="I_SET_BACKTRACE"><code>set_backtrace(errinfo) -&gt; nil | String | [String]</code><span class="permalink">[<a href="../method/Exception/i/set_backtrace.html">permalink</a>][<a href="https://docs.ruby-lang.org/en/2.4.0/Exception.html#method-i-set_backtrace">rdoc</a>]</span></dt>
<dd class="method-description">
<p>
バックトレース情報に errinfo を設定し、設定されたバックトレース情報を返します。
</p>
<dl>
<dt class='method-param'>[PARAM] errinfo:</dt>
<dd>
nil、<a href="../class/String.html">String</a> あるいは <a href="../class/String.html">String</a> の配列のいずれかを指定します。
</dd>
</dl>
<pre class="highlight ruby">
<span class="caption">例:</span>
<code>
<span class="k">begin</span>
  <span class="k">begin</span>
    <span class="nb">raise</span> <span class="s2">"</span><span class="s2">inner</span><span class="s2">"</span>
  <span class="k">rescue</span>
    <span class="nb">raise</span> <span class="s2">"</span><span class="s2">outer</span><span class="s2">"</span>
  <span class="k">end</span>
<span class="k">rescue</span>
  <span class="vg">$!</span><span class="p">.</span><span class="nf">backtrace</span> <span class="c1"># =&gt; [&quot;/path/to/test.rb:5:in `rescue in &lt;main&gt;'&quot;, &quot;/path/to/test.rb:2:in `&lt;main&gt;'&quot;]
</span>  <span class="vg">$!</span><span class="p">.</span><span class="nf">set_backtrace</span><span class="p">(</span><span class="p">[</span><span class="s2">"</span><span class="s2">dummy1</span><span class="s2">"</span>, <span class="s2">"</span><span class="s2">dummy2</span><span class="s2">"</span><span class="p">]</span><span class="p">)</span>
  <span class="vg">$!</span><span class="p">.</span><span class="nf">backtrace</span> <span class="c1"># =&gt; [&quot;dummy1&quot;, &quot;dummy2&quot;]
</span><span class="k">end</span>
</code></pre>
</dd>


</dl>


<div id="footer">
  <a rel="license" href="https://creativecommons.org/licenses/by/3.0/">
    <img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by/3.0/88x31.png" />
  </a>

  フィードバックは<a href="https://github.com/rurema/doctree/issues/new" id="feedback-link">こちら</a>
  <script>if (window.URLSearchParams) { document.getElementById("feedback-link").search = new URLSearchParams({'body': document.location}); }</script></div>
</body>
</html>
